### DISCUSSION TITLE
Is the assignment of ctx->fetched_cipher required even if EVP_CIPHER object used is the same?

### DISCUSSION THREAD

**Discussion by KostasTsiounis** [openssl/openssl#22189] [Q&A]
In our application we are using the `EVP_CipherInit_ex/EVP_CipherUpdate/EVP_CipherFinal_ex` set of methods to perform AES/GCM. After moving to OpenSSL 3.x from 1.1.1, we saw a significant regression in multithreaded performance scenarios. After getting call graph data using `perf`, we found out that a lot of the time is spent in `EVP_CipherInit_ex` -> `EVP_CIPHER_up_ref`, which results in a substantial amount of thread-locking and unlocking.

Looking further into the openssl code, we found this particular part: https://github.com/openssl/openssl/blob/91bc783a93a2a695fe6a2f8da93cf5b5e086ba42/crypto/evp/evp_enc.c#L204

In this part, the `ctx->fetched_cipher` is freed, then a call to `EVP_CIPHER_up_ref` is made and, finally `ctx->fetched_cipher = (EVP_CIPHER *)cipher;`. However, we think that this whole if statement could be avoided if `ctx->fetched_cipher == cipher`.

After changing the condition to `if ((cipher != ctx->fetched_cipher) && (cipher->prov != NULL))`, building and retesting, we saw a 3x improvement when using multiple threads.

The question here is, is there a scenario where even if `cipher` and `ctx->fetched_cipher` point to the same struct, the steps that are part of this if statement still need to be executed?

**Comment by paulidale** [comment]
Doing that would throw off the reference counting I suspect.  The reason for `fetched_cipher` is to distinguish from the legacy path case -- we can't just free `cipher` because it might not be freeable.

I suspect this a symptom not the cause of your issue.  My guess would be implicit fetches.  This [section of the migration guide](https://www.openssl.org/docs/man3.0/man7/migration_guide.html#Algorithm-Fetching) might help.

**Comment by KostasTsiounis** [comment]
We have changed our application code to use explicit fetches, as mentioned in the migration guide, but that didn't fully cover the regression caused by moving to 3.x.

As far as the reference counter goes, if `cipher` and `ctx->fetcher_cipher` point to the same `EVP_CIPHER` struct the code in the if statement mentioned does an `EVP_CIPHER_up_ref`: https://github.com/openssl/openssl/blob/91bc783a93a2a695fe6a2f8da93cf5b5e086ba42/crypto/evp/evp_enc.c#L204 and then an `EVP_CIPHER_free`:https://github.com/openssl/openssl/blob/91bc783a93a2a695fe6a2f8da93cf5b5e086ba42/crypto/evp/evp_enc.c#L208 on the same thing for no reason. I'm only suggesting that we skip it in the particular scenario where `ctx->fetched_cipher == cipher`.

**Comment by mattcaswell** [comment]
> we found out that a lot of the time is spent in EVP_CipherInit_ex -> EVP_CIPHER_up_ref, which results in a substantial amount of thread-locking and unlocking

This suggests to me that OpenSSL is not using atomics on your platform. Ref counter increments/decrements are usually done using atomics instead of locks. I would expect atomics to be available on most modern systems. So, unless you are on some old platform, possibly the atomics detection has failed or we haven't added atomics support yet for your platform. What are you running this on? Did you compile OpenSSL from source?

Also you said 3.x...is that 3.0 or 3.1? There are some significant performance differences between the two.

**Comment by KostasTsiounis** [comment]
We tried all 3.0, 3.1 and even 3.2-alpha. And you are right, there are differences in performance, but none of them covers the performance regression observed when moving from 1.1.1 and using EVP methods.

We are using x86-64 and we are compiling all versions from source. Maybe there is a specific flag for the configuration file we might be missing regarding atomics?

Also, I probably misspoke. When I mention locking, atomics could be part of it. The thing is that this part of the code appears hot in the perf data even when reusing an `EVP_CIPHER_CTX` and not reinitializing the `EVP_CIPHER`, which shouldn't be the case.

Just by adding an extra check in the condition of the if statement (`if ((cipher != ctx->fetched_cipher) && (cipher->prov != NULL))`), the throughput of our application improved by 3x. 

Is this an appropriate fix to avoid unnecessary calls that increase and decrease the reference counter on the same struct? In our use case, we want to reuse the already fetched `EVP_CIPHER`, in which case the code in the if statement is not really needed, at least as far as I understand it.

Here are some perf call graph data before and after that change:

**Before:** 
<img width="1149" alt="multi-GCM-altered-3011-encrypt" src="https://github.com/openssl/openssl/assets/121899919/7486fa6a-7153-464c-87b4-9a3d265de3e1">

**After:**
<img width="994" alt="multi-GCM-altered-320-changed-encrypt" src="https://github.com/openssl/openssl/assets/121899919/2ddbc868-30a7-4986-a771-1aae38a0efcc">

**Comment by t8m** [comment]
It is non-trivial to ensure there are no regressions caused by the optimization you suggest. Would it be instead possible for your application to simply not set the cipher argument (i.e., keep it NULL for the subsequent calls)? It would have the same effect.

**Comment by KostasTsiounis** [comment]
That's actually the problem here. Even if I don't set the `cipher` (i.e., keep it NULL), it is still set here: https://github.com/openssl/openssl/blob/91bc783a93a2a695fe6a2f8da93cf5b5e086ba42/crypto/evp/evp_enc.c#L181, which means that it always goes into that if statement.

**Reply by t8m** [reply]
Ah, right, now I see it. Yeah, that is an obvious hot-path and should be optimized.

**Reply by t8m** [reply]
I'll do a PR

**Reply by KostasTsiounis** [reply]
Sounds great. Please let me know once it's done.

**Reply by KostasTsiounis** [reply]
By the way, this is my fix: https://github.com/KostasTsiounis/openssl/tree/upref_fix

I built the source with it and run and passed all the accompanying tests.

**Reply by t8m** [reply]
Actually the `cipher->prov != NULL` should not be needed here, but I've kept it as an assert in my PR.
